### PR TITLE
*: refactor and fix the usages of DecodeIndexKV for clustered index

### LINKED ISSUE #23106: Adding unique index on duplicate rows hangs (clustered_index + new collation)
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
[config]
new_collations_enabled_on_first_bootstrap = true
```

Remove old data. For example,
```
$ rm /tmp/tidb -r
```

**Note that the clustered index is enabled.**

```sql
drop table if exists t;
create table t (id1 int, id2 varchar(10), a1 int, primary key(id1, id2) clustered) collate utf8mb4_general_ci;
insert into t values (1, 'asd', 1), (1, 'dsa', 1);
alter table t add unique index t_idx(id1, a1);
```

### 2. What did you expect to see? (Required)

```
ERROR 1062 (23000): Duplicate entry '1-1' for key 't_idx'
```

### 3. What did you see instead (Required)

Error message: "number of columns in two handle is different"

```
[INFO] [ddl_worker.go:535] ["[ddl] run DDL job failed, sleeps a while then retries it."] [worker="worker 4, tp add index"] [waitTime=1s] [error="number of columns in two handle is different"]
```

### 4. What is your TiDB version? (Required)

master `2cdd8a30ca99a17fabce6fb888b93aa80b9290a1`

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Problem Summary: Fix #23106.

In the process of the work "remove alter-primary-key configuration", I found `TestAddIndex5` should test the clustered index case but it doesn't because the config `alter-primary-key` is enabled. 

For details, see #23106.

### What is changed and how it works?

What's Changed:

Bugfix:
- checkHandleExists: fix the comparison of the handle.
- (c *indexIter) Next(): fix handle decoding.

Test:
- *: Set the config correctly.

Refactor:
- decodeRestoredValuesV5: reduce the memory allocation, improve the readability.
- BuildRowcodecColInfoForIndexColumns: extract the process of building []rowcodec.ColumnInfo from index columns into a separated function.

### Related changes

NA

### Check List 

Tests 

- Unit test
- Integration test

Side effects

NA

### Release note 

- No release note

### CODE REVIEW

File: table/tables/index.go [hunk 1]
```diff
@@ -351,3 +357,45 @@ func IsIndexWritable(idx table.Index) bool {
...
+	for _, idxCol := range idxInfo.Columns {
+		col := tblInfo.Columns[idxCol.Offset]
+		tps = append(tps, rowcodec.FieldTypeFromModelColumn(col))
+	}
+	return tps
+}
+
+// TryAppendCommonHandleRowcodecColInfos tries to append common handle columns to `colInfo`.
+func TryAppendCommonHandleRowcodecColInfos(colInfo []rowcodec.ColInfo, tblInfo *model.TableInfo) []rowcodec.ColInfo {
+	if !tblInfo.IsCommonHandle {
```
Comment (wjhuang2016) [inline table/tables/index.go:388]: If the version is 0. we don't need it too.

File: session/clustered_index_test.go [hunk 2]
```diff
@@ -381,3 +381,28 @@ func (s *testClusteredSerialSuite) TestClusteredIndexSyntax(c *C) {
...
+	tk.MustExec("insert into t values (1, 'asd', 1), (1, 'dsa', 1);")
+	tk.MustGetErrCode("alter table t add unique index t_idx(id1, a1);", errno.ErrDupEntry)
+
+	tk.MustExec("drop table if exists t;")
+	tk.MustExec("create table t (id1 int, id2 varchar(10), a1 int, primary key(id1, id2) clustered, unique key t_idx(id1, a1)) collate utf8mb4_general_ci;")
+	tk.MustExec("begin;")
+	tk.MustExec("insert into t values (1, 'asd', 1);")
+	tk.MustQuery("select * from t use index (t_idx);").Check(testkit.Rows("1 asd 1"))
+	tk.MustExec("commit;")
+	tk.MustExec("admin check table t;")
```
Comment (wjhuang2016) [inline session/clustered_index_test.go:407]: We may need to drop the table explicitly.

**Comment by tangenta** [comment]
Blocked by https://github.com/tikv/tikv/pull/9182

File: ddl/db_integration_test.go [hunk 3]
```diff
@@ -2225,7 +2225,7 @@ func (s *testIntegrationSuite7) TestAddExpressionIndex(c *C) {
 }
 
 func (s *testIntegrationSuite7) TestCreateExpressionIndexError(c *C) {
-	defer config.RestoreFunc()
+	defer config.RestoreFunc()()
```
Comment (ngaut) [inline ddl/db_integration_test.go:2228]: It is easy to make mistake to call a function of function, is there anyway we can do better?

Comment (tangenta) [reply ddl/db_integration_test.go:2228]: Yeah, it is easy to go wrong. How about this:

```go
restoreFn := config.UpdateGlobal(func(conf *config.Config) {
	conf.Experimental.AllowsExpressionIndex = true
})
defer restoreFn()
```

**Review by AilinKid — COMMENTED** [review]
/lgtm

**Comment by tangenta** [comment]
```
[2021-03-09T11:04:53.261Z] FAIL: 2pc_test.go:735: testCommitterSuite.TestElapsedTTL
[2021-03-09T11:04:53.261Z] 
[2021-03-09T11:04:53.261Z] 2pc_test.go:749:
[2021-03-09T11:04:53.261Z]     c.Assert(lockInfo.LockTtl-atomic.LoadUint64(&ManagedLockTTL), Less, uint64(150))
[2021-03-09T11:04:53.262Z] ... compare_one uint64 = 0x97
[2021-03-09T11:04:53.262Z] ... compare_two uint64 = 0x96
```

/run-unit-test

**Pull request merged**